### PR TITLE
Fixed Code Execution in pm2-deploy

### PR DESCRIPTION
**Pull request by d3m0n-r00t**
### 📊 Metadata *
Fixed ```Remote code execution``` in ```pm2-deploy```

### ⚙️ Description *
```pm2-deploy``` is a module that allows to do pm2 deploy. Documentation: http://pm2.keymetrics.io/docs/usage/deployment/

Affected versions of this package are vulnerable to Command Injection. It is possible to inject arbitrary commands by using malicious ```config``` in the ```spawn()``` function.

### 🐛 Proof of Concept (PoC) *
```javascript
var deployForEnv = require('pm2-deploy').deployForEnv;

var deployConfig = {
  prod: {
    "'; touch HACKED; # user": 'node',
    host: '127.0.0.1',
    ref: 'origin/master',
    repo: 'git@github.com:Asjidkalam/test-repo.git',
    path: '/var/www/test-dev'
  }
};

deployForEnv(deployConfig, 'prod', [], function (err, args) {
  if (err) {
    console.error('Deploy failed:', err.message);
    return console.error(err.stack);
  }
  console.log('Success!');
});
```
pm2-poc

### 🔥 Proof of Fix (PoF) *
pm2-pof

### 👍 User Acceptance Testing (UAT)
Changed ```sh -c echo``` to just ```echo```. I didn't see any difference and the app seems to be working fine. 
Ps: I like open-source contribution.

### CODE REVIEW

**Comment by JamieSlome** [comment]
@Unitech - it would be great to get your thoughts/opinions before we merge this in, cheers! 🍰

**Review by Mik317 — CHANGES_REQUESTED** [review]
Hi @d3m0n-r00t :smile:,
if the command to execute is something like `git add "test title"` it won't work anymore, also the second part of the `piping` is considered as an `argument` of the `echo` command while is a `sh script` which must be executed, so it would be better both avoid `space splitting` and instead `echo` just the first part and then execute the second part, so the `script` avoiding `pipelining` in `spawn` which doesn't allow you to use this type of command manipulation :+1: 

Cheers,
Mik

**Comment by d3m0n-r00t** [comment]
Hi @Mik317, 
Sorry I didn't notice the `piping` at first. When I did a research on piping commands with `spawn`, the second process always exited with ```code 1```. So I updated the fix so that all the arguments are now passed as a list without concatenation.  Hope this would work. Inform me if any problem. 
Thank you

**Review by Mik317 — CHANGES_REQUESTED** [review]
Hi @d3m0n-r00t :smile:
I checked again but it seems to be vulnerable again to the `argument splitting` and since you provide the `sh` command and `-c` argument as first, it doesn't change the code from how it was previously (should be still vulnerable).

I think you should use more calls and save output of the first to avoid using the `sh` in `spawn` and therefore avoid any vulnerability or misfunctionality like `arg split` :+1: 

Cheers,
Mik

**Comment by d3m0n-r00t** [comment]
Hi @Mik317,
Previously the command was given as a ```single``` element to spawn thus was vulnerable. But now every arguments is passed as an array. I think this could avoid ```RCE```. It works for me. 
Screenshot 2020-12-18 223557

Please verify and inform me if I am wrong.
Thank you

File: deploy.js [hunk 1]
```diff
@@ -37,8 +37,8 @@ function spawn(config, args, cb) {
     }).join(' ');
     cmd = [cmd, cmdArgs].join(' ');
   }
-
-  var proc = child_process.spawn('sh', ['-c', cmd], { stdio: 'inherit' });
+  cmd = cmd.split(' ');
```
Comment (Mik317) [inline deploy.js:40]: Splitting arguments will still lead to `unfunctionalities` like when a `dir` like `"test test"` is supplied

**Review by Mik317 — CHANGES_REQUESTED** [review]

**Comment by JamieSlome** [comment]
@mzfr @mufeedvh @bbeale - would be good to get your thoughts here!

**Comment by JamieSlome** [comment]
@d3m0n-r00t - closing this as it is going stale.

Feel free to open a new PR.

Thanks! 🍰

**Comment by huntr-helper** [comment]
Sorry d3m0n-r00t, we enjoyed reviewing your fix but it has not been selected this time. If this bounty has not been closed, please feel free to try again with a new pull request!

We appreciate your effort and look forward to reviewing more of your fixes in the future! 🔨😎

[Come join us on Discord](https://huntr.dev/discord)

**Pull request closed without merging**